lib/puma/cluster.rb

```diff
@@ -181,7 +181,7 @@ def check_workers
         # we need to phase any workers out (which will restart
         # in the right phase).
         #
-        w = @workers.find { |x| x.phase != @phase }
+        w = @workers.find { |x| x.phase < @phase }
```

Comment:
This fixes a race condition bug that causes the tests I'm adding here to fail intermittently, both on master and this branch, but more frequently on the latter. `#check_workers` gets called in the `Cluster#run` loop, and when reforking, there could be a point in time where this method gets called between when worker 0's phase has been incremented, and the phase has been incremented. This results in worker 0 being returned by this find and causes it to receive a SIGTERM below just before a refork happens, which is not ideal. Seeing as that's a special case where the worker's phase could be greater than the target phase changing this ensures that we don't terminate worker 0 in that case while also ensuring workers are incremented to the right phase in an ordinary phased restart scenario.

Reply:
Alternatively we could ensure worker 0's phase is incremented right before the cluster phase is incremented.

Reply:
I could extract this to a separate PR, but it's difficult to test anyway. See the past CI run failures for example.

Reply:
> It would be good to fix this in its own PR, for better documentation of it. Sounds especially useful to me if it is hard to write tests for it.

Can do 👍🏽 

> > See the past CI run failures for example.

> From this branch/PR or have we seen them from main too? Anyway, I think such logs are gone now.

From this branch, but running the test locally on main I was able to replicate the issue.

Reply:
Extracted into https://github.com/puma/puma/pull/3690